### PR TITLE
Manage Catalogue Solutions dashboard

### PR DESCRIPTION
**Pull request by kiran-mastek**
Display all Catalogue Solutions ordered by date created and then by name
Filter functionality
Update admin home link for Catalogue Solutions
Correct condition for Additional Services in Catalogue Item

### CODE REVIEW

File: src/NHSD.GPIT.BuyingCatalogue.Framework/Extensions/EnumExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+                return null;
+
+            FieldInfo field = type.GetField(name);
+            if (field == null)
+                return null;
+
+            return Attribute.GetCustomAttribute(field, typeof(DisplayAttribute)) is DisplayAttribute attr
+                ? attr.Name
+                : value.ToString();
+        }
```
Comment (bjss-chris) [inline src/NHSD.GPIT.BuyingCatalogue.Framework/Extensions/EnumExtensions.cs:23]: This isn't needed – we can add the `Enums.NET` package to the project and use `.AsString(EnumFormat.DisplayName)`.

File: src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Admin/Controllers/CatalogueSolutionsController.cs [hunk 2]
```diff
@@ -0,0 +1,47 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
```
Comment (bjss-chris) [inline src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Admin/Controllers/CatalogueSolutionsController.cs:3]: Are these `using` statements needed?

File: src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Admin/Controllers/CatalogueSolutionsController.cs [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+using NHSD.GPIT.BuyingCatalogue.EntityFramework.Models.GPITBuyingCatalogue;
+using NHSD.GPIT.BuyingCatalogue.ServiceContracts.Solutions;
+using NHSD.GPIT.BuyingCatalogue.WebApp.Areas.Admin.Models;
+
+namespace NHSD.GPIT.BuyingCatalogue.WebApp.Areas.Admin.Controllers
+{
+    [Authorize(Policy = "AdminOnly")]
+    [Area("Admin")]
+    [Route("admin/catalogue-solutions")]
+    public class CatalogueSolutionsController : Controller
```
Comment (bjss-chris) [inline src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Admin/Controllers/CatalogueSolutionsController.cs:16]: `sealed`.

File: src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Admin/Models/CatalogueSolutionsModel.cs [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+using NHSD.GPIT.BuyingCatalogue.EntityFramework.Models.GPITBuyingCatalogue;
+using NHSD.GPIT.BuyingCatalogue.Framework.Extensions;
+
+namespace NHSD.GPIT.BuyingCatalogue.WebApp.Areas.Admin.Models
+{
+    public sealed class CatalogueSolutionsModel
+    {
+        private readonly List<PublicationStatusModel> publicationStatusModels = Enum.GetValues<PublicationStatus>()
+            .ToList()
+            .Select(p => new PublicationStatusModel { Id = (int)p, Display = p.GetDisplayName(), })
```
Comment (bjss-chris) [inline src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Admin/Models/CatalogueSolutionsModel.cs:13]: Trailing comma not needed here.

File: src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Admin/Models/CatalogueSolutionsModel.cs [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+            .ToList();
+
+        public CatalogueSolutionsModel(IList<CatalogueItem> solutions)
```
Comment (bjss-chris) [inline src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Admin/Models/CatalogueSolutionsModel.cs:16]: Parameter can be of type `IEnumerable<CatalogueItem>`?

File: src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Admin/Models/PublicationStatusModel.cs [hunk 6]
```diff
@@ -0,0 +1,11 @@
+﻿namespace NHSD.GPIT.BuyingCatalogue.WebApp.Areas.Admin.Models
+{
+    public class PublicationStatusModel
```
Comment (bjss-chris) [inline src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Admin/Models/PublicationStatusModel.cs:3]: `sealed`

File: src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Admin/Models/PublicationStatusModel.cs [hunk 7]
```diff
@@ -0,0 +1,11 @@
...
+    {
+        public int Id { get; set; }
```
Comment (bjss-chris) [inline src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Admin/Models/PublicationStatusModel.cs:5]: Can this be of type `PublicationStatus`?

File: src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Admin/Views/CatalogueSolutions/Index.cshtml [hunk 8]
```diff
@@ -0,0 +1,136 @@
+﻿@using NHSD.GPIT.BuyingCatalogue.WebApp.Areas.Admin.Models
```
Comment (bjss-chris) [inline src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Admin/Views/CatalogueSolutions/Index.cshtml:1]: Is this `using` needed?

File: src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Admin/Views/CatalogueSolutions/Index.cshtml [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+                        Apply filter
+                    </button>
+                </form>
+                <a asp-action="Index" asp-controller="CatalogueSolutions">Clear filter</a>
+            </div>
+        </details>
+
+        <table role="table" class="nhsuk-table-responsive">
+            <thead role="rowgroup" class="nhsuk-table__head">
+            <tr role="row">
```
Comment (bjss-chris) [inline src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Admin/Views/CatalogueSolutions/Index.cshtml:74]: Can we not use the table component here?

File: src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Admin/Views/Shared/_AdminHeader.cshtml [hunk 10]
```diff
@@ -97,3 +97,12 @@
         </div>
     </nav>
 </header>
+<nav class="nhsuk-breadcrumb" aria-label="Breadcrumb">
+    <div class="nhsuk-width-container">
```
Comment (bjss-chris) [inline src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Admin/Views/Shared/_AdminHeader.cshtml:101]: Can we not use the breadcrumb component here?

File: tests/NHSD.GPIT.BuyingCatalogue.WebApp.UnitTests/Areas/Admin/Controllers/CatalogueSolutionsControllerTests.cs [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+
+        [Theory]
+        [CommonAutoData]
+        public static async Task Post_Index_StatusInput_SetsSelectedOnModel_ReturnsViewWithModel(
+            List<CatalogueItem> solutions)
+        {
+            Enum.GetValues<PublicationStatus>().ToList()
+                .ForEach(
+                    async status =>
+                    {
```
Comment (bjss-chris) [inline tests/NHSD.GPIT.BuyingCatalogue.WebApp.UnitTests/Areas/Admin/Controllers/CatalogueSolutionsControllerTests.cs:68]: I can't see any `await` statements in the test method (only the lambda). Using `async` with `.ForEach` also looks a bit iffy to me.

It feels like this should be a data driven test with each enum value passed in via a parameter.

**Review by bjss-chris — CHANGES_REQUESTED** [review]

File: src/NHSD.GPIT.BuyingCatalogue.Framework/Extensions/EnumExtensions.cs (hunk 1 above)
Comment (kiran-mastek) [reply src/NHSD.GPIT.BuyingCatalogue.Framework/Extensions/EnumExtensions.cs:23]: replaced

File: src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Admin/Controllers/CatalogueSolutionsController.cs (hunk 3 above)
Comment (kiran-mastek) [reply src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Admin/Controllers/CatalogueSolutionsController.cs:16]: done

File: src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Admin/Models/CatalogueSolutionsModel.cs (hunk 5 above)
Comment (kiran-mastek) [reply src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Admin/Models/CatalogueSolutionsModel.cs:16]: changed

File: src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Admin/Models/CatalogueSolutionsModel.cs (hunk 4 above)
Comment (kiran-mastek) [reply src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Admin/Models/CatalogueSolutionsModel.cs:13]: removed

File: src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Admin/Models/PublicationStatusModel.cs (hunk 6 above)
Comment (kiran-mastek) [reply src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Admin/Models/PublicationStatusModel.cs:3]: changed

File: src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Admin/Models/PublicationStatusModel.cs (hunk 7 above)
Comment (kiran-mastek) [reply src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Admin/Models/PublicationStatusModel.cs:5]: renamed

File: src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Admin/Views/CatalogueSolutions/Index.cshtml (hunk 8 above)
Comment (kiran-mastek) [reply src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Admin/Views/CatalogueSolutions/Index.cshtml:1]: removed

File: src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Admin/Views/Shared/_AdminHeader.cshtml (hunk 10 above)
Comment (kiran-mastek) [reply src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Admin/Views/Shared/_AdminHeader.cshtml:101]: Not a good idea as it doesn't use anchor taghelper which allows using `Controller` and `Action` so the URL does not need to be changed if routing changes.

File: src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Admin/Controllers/CatalogueSolutionsController.cs (hunk 2 above)
Comment (kiran-mastek) [reply src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Admin/Controllers/CatalogueSolutionsController.cs:3]: removed

File: src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Admin/Views/CatalogueSolutions/Index.cshtml (hunk 9 above)
Comment (kiran-mastek) [reply src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Admin/Views/CatalogueSolutions/Index.cshtml:74]: no table component, used the tag helper

File: tests/NHSD.GPIT.BuyingCatalogue.WebApp.UnitTests/Areas/Admin/Controllers/CatalogueSolutionsControllerTests.cs (hunk 11 above)
Comment (kiran-mastek) [reply tests/NHSD.GPIT.BuyingCatalogue.WebApp.UnitTests/Areas/Admin/Controllers/CatalogueSolutionsControllerTests.cs:68]: Removed async on the test.
To my knowledge `CommonAutoData` cannot be used along with defined values for the `PublicationStatus` enum in `InlineData`, hence the `.ForEach`.

Comment (bjss-chris) [reply tests/NHSD.GPIT.BuyingCatalogue.WebApp.UnitTests/Areas/Admin/Controllers/CatalogueSolutionsControllerTests.cs:69]: You're correct, you won't be able to use `CommonAutoData`. However, you could use `CommonInlineAutoData` or the `MemberAutoData` attributes, with the latter being the better option in this case I think.

Comment (kiran-mastek) [reply tests/NHSD.GPIT.BuyingCatalogue.WebApp.UnitTests/Areas/Admin/Controllers/CatalogueSolutionsControllerTests.cs]: Spent nearly an hour trying to use attributes suggested to inject `PublicationStatus` and list of `CatalogueItem` but not successful. Issues with unit testing indicate issues with code - in this case coupling requiring injecting domain models into a unit test for controller action. 
Removed the 'iffy' bit with random selection of enum as it doesn't impact the method under test.

File: src/NHSD.GPIT.BuyingCatalogue.EntityFramework/Models/GPITBuyingCatalogue/PublicationStatus.cs [hunk 12]
```diff
@@ -1,10 +1,17 @@
...
     {
+        [Display(Name = "Suspended")]
         Draft = 1,
+
         Unpublished = 2,
+
         Published = 3,
+
+        [Display(Name = "Deleted")]
         Withdrawn = 4,
```
[the remark this answers is not in the document]
Comment (bjss-chris) [reply src/NHSD.GPIT.BuyingCatalogue.EntityFramework/Models/GPITBuyingCatalogue/PublicationStatus.cs]: I think Seana should have an update on the statuses today so we hopefully get this merged tomorrow.

[the remark this answers is not in the document]
Comment (bjss-chris) [reply src/NHSD.GPIT.BuyingCatalogue.EntityFramework/Models/GPITBuyingCatalogue/PublicationStatus.cs]: Based on the updated requirements, `Draft`, `Unpublished` and `Published` can remain the same. We can rename `Withdrawn` to `Suspended` and add a new status `InRemediation`.

**Review by bjss-chris — DISMISSED** [review]

File: src/NHSD.GPIT.BuyingCatalogue.EntityFramework/Models/GPITBuyingCatalogue/PublicationStatus.cs (hunk 12 above)
[the remark this answers is not in the document]
Comment (kiran-mastek) [reply src/NHSD.GPIT.BuyingCatalogue.EntityFramework/Models/GPITBuyingCatalogue/PublicationStatus.cs]: done

**Review by bjss-chris — APPROVED** [review]

**Pull request merged**